source/server/server.cc

```diff
@@ -150,25 +150,34 @@ void InstanceImpl::getParentStats(HotRestart::GetParentStatsInfo& info) {
 
 bool InstanceImpl::healthCheckFailed() { return server_stats_->live_.value() == 0; }
 
-void InstanceUtil::loadBootstrapConfig(envoy::config::bootstrap::v2::Bootstrap& bootstrap,
-                                       const std::string& config_path, bool v2_only) {
-  bool v2_config_loaded = false;
+InstanceUtil::BootstrapVersion
+InstanceUtil::loadBootstrapConfig(envoy::config::bootstrap::v2::Bootstrap& bootstrap,
+                                  Options& options) {
   try {
-    MessageUtil::loadFromFileAndValidate(config_path, bootstrap);
-    v2_config_loaded = true;
+    if (!options.configPath().empty()) {
+      MessageUtil::loadFromFile(options.configPath(), bootstrap);
+    }
+    if (!options.configYaml().empty()) {
+      envoy::config::bootstrap::v2::Bootstrap bootstrap_override;
+      MessageUtil::loadFromYaml(options.configYaml(), bootstrap_override);
+      bootstrap.MergeFrom(bootstrap_override);
+    }
+    MessageUtil::validate(bootstrap);
+    return BootstrapVersion::V2;
   } catch (const EnvoyException& e) {
-    if (v2_only) {
+    if (options.v2ConfigOnly()) {
       throw;
-    } else {
-      // TODO(htuch): When v1 is deprecated, make this a warning encouraging config upgrade.
-      ENVOY_LOG(debug, "Unable to initialize config as v2, will retry as v1: {}", e.what());
     }
+    // TODO(htuch): When v1 is deprecated, make this a warning encouraging config upgrade.
+    ENVOY_LOG(debug, "Unable to initialize config as v2, will retry as v1: {}", e.what());
   }
-  if (!v2_config_loaded) {
-    Json::ObjectSharedPtr config_json = Json::Factory::loadFromFile(config_path);
-    Config::BootstrapJson::translateBootstrap(*config_json, bootstrap);
-    MessageUtil::validate(bootstrap);
+  if (!options.configYaml().empty()) {
+    throw EnvoyException("V1 config (detected) with --config-yaml is not supported");
```

Comment:
Is this case tested? Can we do an EXPECT_THROW_WITH_MESSAGE test here?

Reply:
Done.